cmd/kube-controller-manager/app/storageversionmigrator.go

```diff
@@ -0,0 +1,76 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package app
+
+import (
+	"context"
+
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/metadata"
+	"k8s.io/controller-manager/controller"
+	"k8s.io/kubernetes/pkg/controller/garbagecollector"
+
+	svm "k8s.io/kubernetes/pkg/controller/storageversionmigrator"
+)
+
+const (
+	ResourceVersionControllerName = "resource-version-controller"
+)
+
+func startSVMController(
+	ctx context.Context,
+	controllerContext ControllerContext,
+	controllerName string,
+	garbageCollector *garbagecollector.GarbageCollector,
+) (controller.Interface, bool, error) {
+	config := controllerContext.ClientBuilder.ConfigOrDie(controllerName)
+	client := controllerContext.ClientBuilder.ClientOrDie(controllerName)
+	informer := controllerContext.InformerFactory.Storagemigration().V1alpha1().StorageVersionMigrations()
+
+	dynamicClient, err := dynamic.NewForConfig(config)
+	if err != nil {
+		return nil, false, err
+	}
+
+	discoveryClient, err := discovery.NewDiscoveryClientForConfig(config)
+	if err != nil {
+		return nil, false, err
+	}
+
+	go svm.NewResourceVersionController(
+		ctx,
+		client,
+		discoveryClient,
+		metadata.NewForConfigOrDie(config),
+		informer,
+		controllerContext.RESTMapper,
+		ResourceVersionControllerName,
+	).Run(ctx)
+
+	go svm.NewSVMController(
+		ctx,
+		client,
+		dynamicClient,
+		informer,
+		controllerName,
+		garbageCollector,
+		controllerContext.RESTMapper,
+	).Run(ctx)
+
+	return nil, true, nil
```

Comment:
I'm betting this doesn't (currently) return a controller.Interface because it's nested, don't forget to update after extraction.